### PR TITLE
DISTINCT in order: perf improvement

### PR DESCRIPTION
**Pull request by devcrafter**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
`DISTINCT` in order with `ORDER BY` improves memory usage (significantly) and query execution time if `DISTINCT` columns match (or form a prefix of) `ORDER BY` columns

### CODE REVIEW

**Comment by devcrafter** [comment]
See performance test results on `github_events` dataset [here](https://github.com/ClickHouse/ClickHouse/issues/39313#issuecomment-1190786392)

**Comment by devcrafter** [comment]
#39538 will fix that unclear big memory usage when using `DistinctSortedTransform` as final distinct

**Comment by devcrafter** [comment]
@CurtizJ Please could you review it? I think it's ready

File: src/Processors/Transforms/DistinctSortedChunkTransform.h [hunk 1]
```diff
@@ -32,23 +32,27 @@ class DistinctSortedChunkTransform : public ISimpleTransform
         const SizeLimits & output_size_limits_,
         UInt64 limit_hint_,
         const SortDescription & sorted_columns_descr_,
-        const Names & source_columns_);
+        const Names & source_columns_,
+        bool sorted_stream_);
 
-    String getName() const override { return "DistinctSortedChunkTransform"; }
+    String getName() const override { return (!sorted_stream ? "DistinctSortedChunkTransform" : "DistinctSortedStreamTransform"); }
```
Comment (CurtizJ) [inline src/Processors/Transforms/DistinctSortedChunkTransform.h:38]: Looks like `sorted_stream` field` affects only `getName` behaviour. Do we really need it?

File: tests/queries/1_stateful/00174_distinct_in_order.sql [hunk 2]
```diff
@@ -0,0 +1,10 @@
+select '-- check that distinct with and w/o optimization produce the same result';
```
Comment (CurtizJ) [inline tests/queries/1_stateful/00174_distinct_in_order.sql:1]: Also consider to add setting `optimize_distinct_in_order` to https://github.com/ClickHouse/ClickHouse/blob/8e98b7d47dbbcfe88c8d420b35794854e93a6361/tests/clickhouse-test#L398-L450.

**Review by CurtizJ — COMMENTED** [review]
In general LGTM

File: src/Processors/Transforms/DistinctSortedChunkTransform.h (hunk 1 above)
Comment (devcrafter) [reply src/Processors/Transforms/DistinctSortedChunkTransform.h:38]: Yes, it affects only the name. `DistinctSortedStream` just reflects better what is doing when instantiated for final distinct

File: tests/queries/1_stateful/00174_distinct_in_order.sql (hunk 2 above)
Comment (devcrafter) [reply tests/queries/1_stateful/00174_distinct_in_order.sql:1]: done

**Review by CurtizJ — APPROVED** [review]

File: src/Processors/QueryPlan/DistinctStep.cpp [hunk 3]
```diff
@@ -91,7 +112,11 @@ void DistinctStep::transformPipeline(QueryPipelineBuilder & pipeline, const Buil
 
     if (optimize_distinct_in_order)
     {
+        LOG_DEBUG(getLogger(), "Input sort description    ({}): {}", input_stream.sort_description.size(), dumpSortDescription(input_stream.sort_description));
```
Comment (azat) [inline src/Processors/QueryPlan/DistinctStep.cpp:115]: This looks too verbose, maybe `LOG_TRACE` or even better `LOG_TEST`? @devcrafter what do you think?

Comment (devcrafter) [reply src/Processors/QueryPlan/DistinctStep.cpp:115]: Yes, thanks for feedback. Alexey mentioned it as well. See fix #39955

**Pull request merged**